e2e/whisper/whisper_mailbox_test.go

```diff
@@ -231,10 +230,395 @@ func (s *WhisperMailboxSuite) TestRequestMessageFromMailboxAsync() {
+			"params":["` + bobGroupChatData.Key + `"],
+			"id":1}`)
+	symkeyAddResp = returnedIDResponse{}
+	err = json.Unmarshal([]byte(resp), &symkeyAddResp)
+	s.Require().NoError(err)
+	charlieGroupChatSymkeyID := symkeyAddResp.Result
+	s.Require().NotEmpty(charlieGroupChatSymkeyID)
+
+	//3. charlie create message filter to node by group chat topic
+	resp = charlieRpcClient.CallRaw(`{
+			"jsonrpc": "2.0",
+			"method": "shh_newMessageFilter", "params": [
+				{"symKeyID": "` + charlieGroupChatSymkeyID + `", "topics": [ "` + charlieGroupChatData.Topic + `"], "allowP2P":true}
+			],
+			"id": 1
+		}`)
+
+	msgFilterResp = returnedIDResponse{}
+	err = json.Unmarshal([]byte(resp), &msgFilterResp)
+	s.Require().NoError(err)
+	charlieGroupChatMessageFilterID := msgFilterResp.Result
+
+	helloWorldMessage := hexutil.Encode([]byte("Hello world!"))
+	//Post message
+	resp = aliceRpcClient.CallRaw(`{
+		"jsonrpc": "2.0",
+		"method": "shh_post",
+		"params": [
+			{
+			"symKeyID": "` + groupChatKeyID + `",
+			"topic": "` + groupChatTopic.String() + `",
+			"payload": "` + helloWorldMessage + `",
+			"powTarget": 0.001,
+			"powTime": 2
+			}
+		],
+		"id": 1}`)
+	postResp = baseRPCResponse{}
+	err = json.Unmarshal([]byte(resp), &postResp)
+	s.Require().NoError(err)
+	s.Require().Nil(postResp.Err)
+	time.Sleep(time.Second)
+
+	resp = bobRpcClient.CallRaw(`{
+		"jsonrpc": "2.0",
+		"method": "shh_getFilterMessages",
+		"params": ["` + bobGroupChatMessageFilterID + `"],
+		"id": 1}`)
+	messages = getFilterMessagesResponse{}
+	err = json.Unmarshal([]byte(resp), &messages)
+	s.Require().NoError(err)
+	s.Require().Equal(1, len(messages.Result))
+	s.Require().Equal(helloWorldMessage, messages.Result[0]["payload"].(string))
+	s.Require().NoError(err)
+
+	resp = charlieRpcClient.CallRaw(`{
+		"jsonrpc": "2.0",
+		"method": "shh_getFilterMessages",
+		"params": ["` + charlieGroupChatMessageFilterID + `"],
+		"id": 1}`)
+	messages = getFilterMessagesResponse{}
+	err = json.Unmarshal([]byte(resp), &messages)
+	s.Require().NoError(err)
+	s.Require().Equal(1, len(messages.Result))
+	s.Require().Equal(helloWorldMessage, messages.Result[0]["payload"].(string))
+	s.Require().NoError(err)
+
+	resp = bobRpcClient.CallRaw(`{
+		"jsonrpc": "2.0",
+		"method": "shh_getFilterMessages",
+		"params": ["` + bobGroupChatMessageFilterID + `"],
+		"id": 1}`)
+	messages = getFilterMessagesResponse{}
+	err = json.Unmarshal([]byte(resp), &messages)
+	s.Require().NoError(err)
+	s.Require().Equal(0, len(messages.Result))
+	s.Require().NoError(err)
+
+	resp = charlieRpcClient.CallRaw(`{
+		"jsonrpc": "2.0",
+		"method": "shh_getFilterMessages",
+		"params": ["` + charlieGroupChatMessageFilterID + `"],
+		"id": 1}`)
+	messages = getFilterMessagesResponse{}
+	err = json.Unmarshal([]byte(resp), &messages)
+	s.Require().NoError(err)
+	s.Require().Equal(0, len(messages.Result))
+
+	//Request each one messages from mailbox using enode
+	resp = bobRpcClient.CallRaw(`{
```

Comment:
Was Bob offline when the messages were sent? How do you make sure that the message was received from the MailServer and not from other nodes?

Reply:
This comment is resolved by calling `shh_getFilterMessages` before requesting messages from MailServer. If `shh_getFilterMessages` returns no messages and after requesting messages `shh_getFilterMessages` returns a new message, this means that message comes from the MailServer.